Solver result caching

m-yac opened the pull request:
This PR adds a solver-backend-aware system for caching SMT solver results.
- Check out the [**Caching Solver Results**](https://github.com/GaloisInc/saw-script/blob/a960c3665db04562d30b9160682bcef359cb5fff/doc/manual/manual.md#caching-solver-results) section of the SAW manual for a description and an example of this feature.
- Check out [the top of `SolverCache.hs`](https://github.com/GaloisInc/saw-script/blob/a960c3665db04562d30b9160682bcef359cb5fff/src/SAWScript/SolverCache.hs#L8) for more information on how the system works internally.

## Example (Out of date, see the manual for an update)

First we enable caching to disk and run a few queries:
```
sawscript> enable_experimental 
sawscript> set_solver_cache_path "test.cache"
sawscript> prove_print w4 {{ \(x:[64]) -> 2 * x == x + x }}
[18:32:12.839] Caching result: 4b6a78359fb7b9cc (What4 v1.3-29-g6c462cd, Z3 version 4.8.7 - 64 bit)
[18:32:12.840] Theorem (let { x@1 = Vec 64 Bool
      x@2 = TCNum 64
      x@3 = PRingSeqBool x@2
    }
 in (x : x@1)
-> EqTrue
     (ecEq x@1 (PEqSeqBool x@2)
        (ecMul x@1 x@3 (ecNumber (TCNum 2) x@1 (PLiteralSeqBool x@2)) x)
        (ecPlus x@1 x@3 x x)))

sawscript> prove_print z3 {{ \(x:[64])(y:[64]) -> y == x }}
[18:32:18.640] Caching result: 91ad8684b52cf5d1 (SBV 9.2, Z3 version 4.8.7 - 64 bit)

Stack trace:
"prove_print" (<stdin>:1:1-1:12)
prove: 1 unsolved subgoal(s)
Invalid: [x = 18446744073709551615, y = 0]
```
Next, we can use `print_solver_cache` to see what was cached (`print_solver_cache ""` will show all entries) and `print_solver_cache_stats` to get statistics about how the solver cache has been used so far this run:
```
sawscript> print_solver_cache "4b6a78359fb7b9cc"
[18:32:30.093] SHA: 4b6a78359fb7b9cc2c6ca3868b07a6e156a0ab17a9d40ef4ec2d5a48174f00cf
[18:32:30.093] - Result: unsat
[18:32:30.093] - Solver: "W4 ->z3"
[18:32:30.093] - Versions: What4 v1.3-29-g6c462cd, Z3 version 4.8.7 - 64 bit

sawscript> print_solver_cache "91ad8684b52cf5d1"
[18:33:19.790] SHA: 91ad8684b52cf5d1cc586cbe37a8896eee2eb3783cdb8d75d1f602ab66e9bcb5
[18:33:19.790] - Result: sat [(0,18446744073709551615),(1,0)]
[18:33:19.790] - Solver: "SBV->Z3"
[18:33:19.790] - Versions: SBV 9.2, Z3 version 4.8.7 - 64 bit

sawscript> print_solver_cache_stats
[18:33:47.962] == Solver result cache statistics ==
[18:33:47.962] - 2 results cached in /.../test.cache
[18:33:47.962] - 2 insertions into the cache so far this run (0 failed attempts)
[18:33:47.962] - 0 usages of cached results so far this run (0 failed attempts)
```
Finally, we can start up `saw` again and verify that these results will be used – in particular, order doesn't matter, local variable names don't matter, but the solver backend does:
```
sawscript> enable_experimental 
sawscript> set_solver_cache_path "test.cache"
sawscript> prove_print z3 {{ \(new1:[64])(new2:[64]) -> new2 == new1 }}
[18:34:56.895] Using cached result: 91ad8684b52cf5d1 (SBV 9.2, Z3 version 4.8.7 - 64 bit)

Stack trace:
"prove_print" (<stdin>:1:1-1:12)
prove: 1 unsolved subgoal(s)
Invalid: [new1 = 18446744073709551615, new2 = 0]

sawscript> prove_print w4 {{ \(new1:[64])(new2:[64]) -> new2 == new1 }}
[18:35:02.952] Caching result: 6ab538591daf0e8f (What4 v1.3-29-g6c462cd, Z3 version 4.8.7 - 64 bit)

Stack trace:
"prove_print" (<stdin>:1:1-1:12)
prove: 1 unsolved subgoal(s)
Invalid: [new1 = 18446744073709551615, new2 = 0]

sawscript> prove_print w4 {{ \(new:[64]) -> 2 * new == new + new }}
[18:35:08.710] Using cached result: 4b6a78359fb7b9cc (What4 v1.3-29-g6c462cd, Z3 version 4.8.7 - 64 bit)
[18:35:08.710] Theorem (let { x@1 = Vec 64 Bool
      x@2 = TCNum 64
      x@3 = PRingSeqBool x@2
    }
 in (new : x@1)
-> EqTrue
     (ecEq x@1 (PEqSeqBool x@2)
        (ecMul x@1 x@3 (ecNumber (TCNum 2) x@1 (PLiteralSeqBool x@2))
           new)
        (ecPlus x@1 x@3 new new)))
```
There are also two other ways to interact with solver caching:
- The `enable_solver_cache` command, which turns on solver result caching in memory but does not save anything to disk. If `set_solver_cache_path` is later run, everything from memory is written to disk. If `set_solver_cache_path` is never run, the cached results are forgotten after `saw` execution ends.
- The `SAW_SOLVER_CACHE_PATH` environment variable, which has the same behavior as calling `set_solver_cache_path` with the value of the variable at the start of execution.

A quick and dirty test on a larger example shows that there is a performance improvement – in the case of `intTests/test_w4/prims.saw` on my machine (see below) it takes 3.467s to run the test with caching turned off, 3.541s to run the test with caching turned on but nothing in the cache (so 0.1s of extra time spent caching the entries), and 2.540s to run the test with a loaded cache (so 1s less time calling solvers, a ~30% speed increase). I imagine for files with more complicated SMT queries than `prims.saw` (which mostly contains zeroth-order queries such as `(16:[8]) >> 3 == 2`), this speed increase can get much larger.
```
$ time ./bin/saw intTests/test_w4/prims.saw > /dev/null
./bin/saw intTests/test_w4/prims.saw > /dev/null  2.75s user 0.58s system 96% cpu 3.467 total
$ time SAW_SOLVER_CACHE_PATH="prims.cache" ./bin/saw intTests/test_w4/prims.saw > /dev/null
SAW_SOLVER_CACHE_PATH="prims.cache" ./bin/saw intTests/test_w4/prims.saw >   2.74s user 0.57s system 93% cpu 3.541 total
$ time SAW_SOLVER_CACHE_PATH="prims.cache" ./bin/saw intTests/test_w4/prims.saw > /dev/null
SAW_SOLVER_CACHE_PATH="prims.cache" ./bin/saw intTests/test_w4/prims.saw >   1.99s user 0.43s system 95% cpu 2.540 total
```

## Additional Changes

Note that this PR also makes the following changes:
- The `Setup.hs` script now also includes AIG, What4, and SBV version information in `GitRev.hs`, used by `Builtins.hs` and `Interpreter.hs` to generate solver backend version information for caching. Accordingly the `saw-script` library and executable now also depend on `GitRev` in `saw-script.cabal`.
- Instead of every solver backend individually calling `sequentToSATQuery` on its given `Sequent`, we now call `sequentToSATQuery` in one place (`applyProverToGoal`) and pass the resulting `SATQuery` to each backend directly.

bboston7 commented:
> @bboston7 has some thoughts on this.

To clarify for the curious: I used sqlite extensively at my previous job and from experience it gets a bit tricky when concurrency is involved. You'll notice that [the page on when to use sqlite](https://www.sqlite.org/whentouse.html) calls out high levels of concurrent writes as something sqlite does not handle well.  In practice there are tricks to make this better (such as enabling [WAL mode](https://www.sqlite.org/wal.html)), but still any application that makes use of concurrent writes (or reads while writing) will need to handle sqlite errors that can arise at any point a query is made. Usually a simple "wait and try again" approach works, but ultimately it can take a few months of tweaks to get everything working well.

Additionally, sqlite's performance can vary significantly depending on how you structure your tables and queries. We would probably want to take the time to design that out before implementing it.

It's not that I think we shouldn't use a database for this at some point, I just wanted to push back on the idea that it's a quick change to use sqlite in a concurrent application like this.

kquick commented:
Can you say more about that What4 cache file that is next to the Z3 cache file?

What4 tests currently do determine solver versions; it seems reasonable that this could be moved into the what4 library itself and surfaced somewhere in the API: https://github.com/GaloisInc/what4/blob/master/what4/test/ProbeSolvers.hs#L39

My opinion is that we definitely want a solver-specific cache (different solvers can give different answers) and a version-specific solver cache (does a new version of a solver still give the same answers?).

m-yac commented:
> Can you say more about that What4 cache file that is next to the Z3 cache file?

Not sure exactly what you're asking, but currently the cache files all have the same format: the filename is the hash of the saw-core term which is passed to the solver backend, and the solver version is just saved as a string in that file.

> What4 tests currently do determine solver versions; it seems reasonable that this could be moved into the what4 library itself and surfaced somewhere in the API: https://github.com/GaloisInc/what4/blob/master/what4/test/ProbeSolvers.hs#L39

Amazing - thanks for the link! I'm going to use this to quickly get a version where we also cache by solver version up and running and I'll update the example.

kquick commented:
> > Can you say more about that What4 cache file that is next to the Z3 cache file?
> 
> Not sure exactly what you're asking, but currently the cache files all have the same format: the filename is the hash of the saw-core term which is passed to the solver backend, and the solver version is just saved as a string in that file.

Oh, I misread the original and thought you only had one query, but you generated two different queries from the `sawscript>` prompt.  It's clear to me now.

eddywestbrook commented:
Let's be clear about goals: the primary goal of this SMT caching feature is to speed up CI. It could maybe be useful in other contexts, but CI is the main use case of interest.

Now, the easiest way to use this sort of cache in github CI is to use the github cache action, which gives each workflow its own local version of the cache. There doesn't seem to be a straightforward way to get different workflows to share a cache, and in fact different workflows will probably make different queries anyway on the whole, so separating the caches to different CI workflows is probably a good thing in the end. Thus, we only really have to worry about concurrency inside a single workflow. Some of our existing workflows are completely sequential, since they are just a single SAW file. But things like the `intTests` run a bunch of different tests, which at least should be run concurrently, though I'm not sure if they currently are.

Given all of this, I would say we only need to support a low degree of concurrency. SQLite seems feasible, but OTOH the one file per query approach seems intriguing as well, for the reasons you outlined. And in fact, the one file per query approach could actually have better performance, because most of the concurrent processes in a CI run will be making different SMT queries: the one file per query approach would allow these to happen concurrently, while SQLite would serialize the writes. It should even be straightforward to correctly handle the concurrency here: the `O_CREAT` plus `O_EXCL` flags to the `open` syscall should make file creation an atomic action, and the SMT cache could be implemented so that only the process that creates the cache file gets to write to it. A reader could still see an SMT cache file that has been only partially written, but the failure case should be either to try again or, as @m-yac says above, to just do the query anyway. So maybe the existing one file per SMT query approach is the way to go. And, as @m-yac says, it's the easiest, so maybe we should just try it for now and see how it fares in the CI system.

My suggestion for SMT solver versions is this: each SMT cache should be globally marked when it is created with the versions of all the solvers it was created with. Whenever a CI workflow is started and the SMT cache is loaded / installed, some process should check those SMT solver versions against the current versions that are installed, and just wipe the cache if any of those solver versions have changed. The expectation is that we are only occasionally going to change SMT solver versions, and when we do, we don't expect to go back to an old version, so there is no reason to keep the old cached values around. I could imagine trying to only wipe the cached queries corresponding to the particular solver whose version was updated, but then you have to associate each query with the solver it was made against, and this creates additionally complexity. Again, we are only occasionally going to change SMT solver versions for any of our solvers, so we might as well save ourselves the trouble and just wipe the whole SMT cache.

Then, for each copy of SAW that gets run in the workflow, the SMT caching component(s) will check if the solver versions it sees are the same as those given in the SMT cache. If not, then something is wrong with the CI setup. It should probably raise an error, though I could imagine cases where we want to turn that into a warning instead, though it should be a stern warning. :)  If it's a warning and not an error, it should at least be something that the user can easily see when checking the CI logs.

That is my 2 cents. Which buys a lot of text these days, apparently.

m-yac commented:
Caching is now dependent on the solver used as well as the version of the solver used. Thanks @kquick for the pointer!

kquick commented:
> My suggestion for SMT solver versions is this: each SMT cache should be globally marked when it is created with the versions of all the solvers it was created with.

I'm actually more in favor of including the SMT version in the file (and as part of the input to the hash for the file's name).  I don't see it as a lot of additional work, and while the primary utilization of the cache is intended to be the CI, the caching is being implemented by saw-script itself and not the CI, so I think it should be robust against other uses.  It's also worth noting that the what4 repo tests multiple different solver versions during its CI and that might be desirable for saw at some future point as well.

weaversa commented:
Along with saving/hashing against solver version, I suggest also
saving/hashing against any command line options to the solver, in case they
also change.

Our CI scripts mostly use the Python interfaces though both saw-remote-api
and cryptol-remote-api, so it would be _great_ to give both python
interfaces access to these new caching options.

On Fri, May 19, 2023 at 9:51 PM Kevin Quick ***@***.***>
wrote:

> My suggestion for SMT solver versions is this: each SMT cache should be
> globally marked when it is created with the versions of all the solvers it
> was created with.
>
> I'm actually more in favor of including the SMT version in the file (and
> as part of the input to the hash for the file's name). I don't see it as a
> lot of additional work, and while the primary utilization of the cache is
> intended to be the CI, the caching is being implemented by saw-script
> itself and not the CI, so I think it should be robust against other uses.
> It's also worth noting that the what4 repo tests multiple different solver
> versions during its CI and that might be desirable for saw at some future
> point as well.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/GaloisInc/saw-script/pull/1873#issuecomment-1555421243>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABYQSOTGC5ST3IJQWPZHLPDXHAPSDANCNFSM6AAAAAAYICQ3C4>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

andreistefanescu commented:
I would suggest to use a persistent key/value store rather than a SQL database. I think [LMDB](https://www.symas.com/lmdb) is simple to use and fits well. There are haskell binding, such as [lmdb-simple](https://hackage.haskell.org/package/lmdb-simple). Alternatively, RocksDB (https://rocksdb.org/) supports more features, but I don't think we require any here. There are haskell binding as well, such as [rocksdb-haskell](https://hackage.haskell.org/package/rocksdb-haskell-jprupp).

m-yac commented:
Ha, well guess my change to CI only works on mac (i.e. the only place I tested it).
- On linux it looks like it's `make install` that fails, specifically it doesn't have permissions to add a file to `/usr/local/lib/`
- On Windows it looks like the installation of `lmdb` simply didn't work, since `cabal` doesn't see it when trying to install the Haskell bindings.

I am very out of my depth on CI stuff so any suggestions would be very appreciated...

> Our CI scripts mostly use the Python interfaces though both saw-remote-api and cryptol-remote-api, so it would be _great_ to give both python interfaces access to these new caching options.

Forgot to address this, but yes – bindings for the `saw-remote-api` will definitely be added! Once this is merged a separate set of changes will need to be made to the `cryptol` repo to add solver result caching there, but included in that will be `cryptol-remote-api` bindings.

andreistefanescu commented:
on ubuntu, `apt-get install liblmdb-dev`. On macos, `brew install lmdb`. For ci, this seems to support linux/mcos/windows: https://github.com/input-output-hk/haskell-lmdb

m-yac commented:
Thanks for the suggestion @andreistefanescu! Unfortunately even with the CI script from your link, I'm still getting the following error during `cabal install` on Windows:
```
Error: cabal.exe: Missing dependency on a foreign library:
* Missing (or bad) C library: lmdb
```

Is it possible that [input-output-hk/haskell-lmdb](https://github.com/input-output-hk/haskell-lmdb) only depends on the `bin` of lmdb? I noticed that after the build of lmdb they only add `C:\msys64\mingw64\bin` to the `PATH` (`$env:PATH=("C:\msys64\mingw64\bin;{0}" -f $env:PATH)`). I tried also adding `C:\msys64\mingw64\lib` and `C:\msys64\mingw64\include` to the `PATH`, but that didn't help `cabal` see the library. Is there some other environment variable I need to set for Windows to tell it where to look for C libraries/includes?

andreistefanescu commented:
try: https://github.com/input-output-hk/haskell-lmdb/blob/master/.github/workflows/cabal.project.local.Windows#LL8C1-L9C48
maybe only for `lmdb-simple` if https://github.com/GaloisInc/saw-script/blob/master/.github/workflows/ci.yml#L118-L120 is still an issue

m-yac commented:
Okay, this PR is now ready for review! @RyanGlScott, @kquick, or anyone else – any idea who might be the right person to review this PR?

As an update to the above conversation, we ended up deciding to use the Python LMDB bindings instead of something like `lmdb-simple`, since the latter requires the LMDB dynamically-linked C libraries to be present at runtime (even if you don't use solver caching). Adding this extra requirement for using the `saw` executable was deemed too much, especially given the fact that there is [no official packaged release of LMDB](https://www.openldap.org/faq/data/cache/108.html) and installation on Windows can be difficult. Instead, if and only if the user enables solver caching, they must have Python 3 and [the Python LMDB bindings](https://lmdb.readthedocs.io/en/release/) installed (the latter of which do not depend on any dynamically linked libraries and can be installed via `pip`). We do this by spawning a Python process when the user enables solver caching - see `LMDBOptDatabse.hs` and `lmdb_opt_database.py`.

RyanGlScott commented:
I definitely want to do a careful review of this, but it may take me a bit to read through all of this. This is definitely on my radar, however.

RyanGlScott commented on Setup.hs:
```diff
@@ -15,22 +16,46 @@ myBuild pd lbi uh flags = do
+    ex <- doesFileExist "cabal.project.freeze"
+    unless ex $ callProcess cabal ["freeze"]
+    wss <- fmap words <$> lines <$> readFile "cabal.project.freeze"
+    unless ex $ removeFile "cabal.project.freeze"
+    case find (\ws -> (ws !! 0) == "any.sbv") wss of
+      Just ws -> return . Just . init . drop 2 $ ws !! 1
+      Nothing -> return Nothing
 
   writeFile (dir </> "GitRev.hs") $ unlines
     [ "module GitRev where"
     , "hash :: String"
-    , "hash = " ++ show (init desc)
+    , "hash = " ++ show desc
```
Perhaps it's worth leaving a comment about what kind of `hash` in particular this is, since there are now multiple types of hashes being written here.

RyanGlScott commented on saw-remote-api/src/SAWServer.hs:
```diff
@@ -205,6 +208,12 @@ initialState readFileFn =
      jvmTrans <- CJ.mkInitialJVMContext halloc
      cwd <- getCurrentDirectory
+     mb_cache <- lookupEnv "SAW_SOLVER_CACHE_PATH" >>= \case
+      Just p | length p > 0 -> do
```
I'd phrase this code this way:

```suggestion
      Just p | not (null p) -> do
```

So that this will short-circuit if `p` is an empty string.

RyanGlScott commented on src/SAWScript/Interpreter.hs:
```diff
@@ -452,6 +456,12 @@ buildTopLevelEnv proxy opts =
        jcb <- JCB.loadCodebase (jarList opts) (classPath opts) (javaBinDirs opts)
        currDir <- getCurrentDirectory
+       mb_cache <- lookupEnv "SAW_SOLVER_CACHE_PATH" >>= \case
+        Just p | length p > 0 -> do
```
```suggestion
        Just p | not (null p) -> do
```

RyanGlScott commented on src/SAWScript/Interpreter.hs:
```diff
@@ -1057,6 +1085,52 @@ primitives = Map.fromList
+    , " cache have been made so far this session, and how many times"
+    , " cached results have been used so far this session." ]
+
+  , prim "test_solver_cache_stats" "[Int] -> TopLevel ()"
+    (pureVal (onSolverCache . testSolverCacheStats))
+    Experimental
+    [ "Check whether the values of the statistics printed out by"
+    , " print_solver_cache_stats are equal to those given, failing if"
+    , " this does not hold. Used for testing." ]
```
I'm rather confused as to how you figure out what arguments to pass to `test_solver_cache_stats`. The description for `print_solver_cache_stats` suggests that this command has three different statistics, but every call site to `test_solver_cache_stats` is a list with _six_ arguments. What explains the discrepancy there?

Also, is there a good reason for this to be a list? Why not six separate `Int` arguments?

m-yac replied:
Good point, I have absolutely no idea why I implemented this as a `[Int]`. Fixed, and also added clear docs about what values are printed out by `print_solver_cache_stats` and tested by `test_solver_cache_stats`.

RyanGlScott commented on src/SAWScript/SolverCache/LMDBOptDatabase.hs:
```diff
@@ -0,0 +1,292 @@
+  -- file every time we open a database
+  hPutStr fh lmdb_opt_database__py >> hClose fh
+  (mb_i, mb_o, mb_e, p) <- createProcess $
+    (shell $ "python3 " ++ fnm ++ " shell")
```
Needing to have `python` installed is a new form of dependency, and one that may be surprising if you're not expecting it. I think it would be worth documenting this in the `:?` output for `enable_solver_cache`, as well as somewhere in the SAW manual.

m-yac replied:
Good catch! Added better docs to `enable_solver_cache` and `set_solver_cache_path`, working on the SAW manual now.

RyanGlScott commented on src/SAWScript/SolverCache/lmdb_opt_database.py:
```diff
@@ -0,0 +1,202 @@
+     the result as a hex string"""
+  print(bytes(json.dumps(obj), 'utf-8').hex())
+
+def pHexPair(v1, v2, pretty = False):
+  """Print the given pair of values as a hex strings, if both are ``bytes``
+     objects. For the second element, if ``pretty`` is ``True``, deserialise it
+     as JSON and print that instead."""
```
This `pretty` argument is rather interesting, as it completely changes the behavior of the function. I don't think this is necessarily a bad idea, but I wonder if it's worth having entirely separate functions for printing things as JSON instead of as hex. (Relatedly, I'm not sure where this `pretty` argument is actually used.)

m-yac replied:
This and the other `pretty` were only there for manual debugging while I was writing this file. Removed since they're not used anymore.

RyanGlScott commented on src/SAWScript/SolverCache/lmdb_opt_database.py:
```diff
@@ -0,0 +1,202 @@
+  """Print the given value as a hex string, if it is a ``bytes`` object"""
+  if isinstance(v, bytes): print(v.hex())
+
+def pJSONHex(obj, pretty = False):
+  """Serialize the given object as JSON to get a ``bytes`` object then print
+     the result as a hex string"""
```
Is `pJSONHex`'s `pretty` argument ever used?

RyanGlScott commented on src/SAWScript/SolverCache.hs:
```diff
@@ -0,0 +1,483 @@
+
+-- Solver Backends -------------------------------------------------------------
+
+-- | A datatype representing one of the solver backends available in SAW. Note
+-- that for 'SBV' and 'W4', multiple backends will be used (e.g. 'SBV' with
+-- 'Z3' or 'W4' with 'AIG' and 'ABC').
+data SolverBackend = What4
+                   | SBV
+                   | AIG
+                   | RME
+                   -- External solvers (copied from SBV.Solver)
+                   | ABC
+                   | Boolector
+                   | Bitwuzla
+                   | CVC4
+                   | CVC5
+                   | DReal
+                   | MathSAT
+                   | Yices
+                   | Z3
```
Several of these backends are present in SBV, but not in SAW (e.g., `Bitwuzla`). I'm not sure if that makes a difference in practice, but it is a little weird seeing them here. I wonder if it's worth removing them here.

m-yac replied:
Honestly the main reason I kept them was the convenience of having [`cvtSolver :: SBV.Solver -> SolverBackend`](https://github.com/GaloisInc/saw-script/blob/6a5ec6725ce0b53f41864694b266779f3a3811bb/src/SAWScript/SolverCache.hs#L183) be total. But, it also means that if anyone ever wants to add support for `Bitwuzla` or `DReal`, all they'd have to do is add a SAW command which calls `proveSBV SBV.Bitwuzla`/`SBV.DReal`. Otherwise they'd also have to update this type, the pattern match in [`cvtSolver`](https://github.com/GaloisInc/saw-script/blob/6a5ec6725ce0b53f41864694b266779f3a3811bb/src/SAWScript/SolverCache.hs#L183) , and the pattern match in [`getSolverBackendVersion`](https://github.com/GaloisInc/saw-script/blob/6a5ec6725ce0b53f41864694b266779f3a3811bb/src/SAWScript/SolverVersions.hs#L44).

That all being said, I totally understand removing them. I don't really have a preference, and am happy to do whatever you think is best.

RyanGlScott replied:
> But, it also means that if anyone ever wants to add support for `Bitwuzla` or `DReal`, all they'd have to do is add a SAW command which calls `proveSBV SBV.Bitwuzla`/`SBV.DReal`.

As someone who recently added support for a new SMT solver... I _wish_ it were that straightforward. :)

With the current state of SAW's code, there are quite a few places in the code that you'd need to update if you want to add a new solver. I suppose the argument here is that only listing the `SolverBackend`s that SAW currently supports would add yet another place in the code that you'd need to update, which is fair. I don't have a strong opinion on this, but if you do leave the current definition of `SolverBackend` in place, I think it would be worth putting all of the supported-in-SBV-but-not-supported-in-SAW solvers in a separate part of the definition, separated by a comment.

m-yac replied:
> As someone who recently added support for a new SMT solver... I wish it were that straightforward. :)

Oh dang it! Shows my naivete :)

I decided to leave those extra two in, just because I think it would be very easy for someone to not realize they have to update this file if any additional SBV backends are added. But, I add a bunch of comments making it clear that SAW does not currently support those solvers, and they are only there to make sure the system for keeping track of solver backend versions doesn't silently break if support for them is ever added. How does it look now?

RyanGlScott commented on src/SAWScript/SolverCache.hs:
```suggestion
-- | The key type for 'SolverCache'. Each is a SHA256 hash of 'SATQuery' and
```

RyanGlScott commented on src/SAWScript/SolverCache.hs:
`fmap ... (mapM (mapM ...) ...)`... there is so much nested mapping going on that I struggle to understand what is going on here. I think it would be worth breaking this up into multiple lines (with `do`-notation) to make it easier to understand at a glance.

m-yac replied:
I wasn't sure how to rewrite this in `do`-notation, but instead I tried pulling the confusing jumble of `mapM`s etc. out as a subfunction with an explicit type. Is something like the below more readable now?
```haskell
  SolverCacheValue vs opts solver_name <$> firstsMaybeM (`elemIndex` ecs) cexs
  where firstsMaybeM :: Monad m => (a -> m b) -> Maybe [(a, c)] -> m (Maybe [(b, c)])
        firstsMaybeM = mapM . mapM . firstM
```
I also made the analogous change to `fromSolverCacheValue` and it's definitely better that what it was before: `(fmap (fmap (first (ecs !!))) cexs, solver_name)`.
```haskell
  (firstsMaybe (ecs !!) cexs, solver_name)
  where firstsMaybe :: (a -> b) -> Maybe [(a, c)] -> Maybe [(b, c)]
        firstsMaybe = fmap . fmap . first
```

RyanGlScott replied:
Yes, that's much clearer now.

RyanGlScott commented on src/SAWScript/SolverCache.hs:
This output can be useful, but I wonder if it is too noisy for most users. Is it worth guarding this behind a higher `-v` level?

m-yac replied:
Good idea! I added them to the `Debug` verbosity level.

RyanGlScott reviewed (commented):
I've done a fairly shallow review of some of the code that stood out after a quick read-through. I'll admit that I haven't spent too much time actually looking at the LMDB code that powers the cache, but I am limited by my lack of familiarity with LMDB. If there are parts that you think deserve closer scrutiny, please let me know.

m-yac commented on Setup.hs:
```diff
@@ -15,22 +16,46 @@ myBuild pd lbi uh flags = do
-            Just git -> readProcess "git" ["describe", "--always", "--dirty"] ""
-                        `catch` gitfailure
-            Nothing -> return "<VCS-less build> "
+  sbv_ver <- withExe "." "cabal" Nothing Nothing $ \cabal -> do
+    ex <- doesFileExist "cabal.project.freeze"
+    unless ex $ callProcess cabal ["freeze"]
+    wss <- fmap words <$> lines <$> readFile "cabal.project.freeze"
+    unless ex $ removeFile "cabal.project.freeze"
+    case find (\ws -> (ws !! 0) == "any.sbv") wss of
+      Just ws -> return . Just . init . drop 2 $ ws !! 1
+      Nothing -> return Nothing
```
Oh wow, yes, that is a much better solution!! I never knew about `VERSION_pkg`, only `MIN_VERSION_pkg` – thanks!

m-yac commented on .github/workflows/ci.yml:
```diff
@@ -432,6 +432,10 @@ jobs:
           java-package: jdk
           architecture: x64
 
+      - shell: bash
+        if: "matrix.suite == 'integration_tests'"
+        run: pip install lmdb
+
```
@kquick (or anyone else with CI knowledge) – any idea why this `pip install` might not be carrying over to the run of `test_solver_cache/test.sh`? I had to [add the same line to `test.sh`](https://github.com/GaloisInc/saw-script/blob/6a5ec6725ce0b53f41864694b266779f3a3811bb/intTests/test_solver_cache/test.sh#L7) in order to get the test to pass on CI.

andreistefanescu commented:
I'm late with this feedback, but I think `LLDB` should be linked statically, not dynamically. Requesting our users to install the library locally, or in CI, is a problem, and we can solve it for them. @m-yac I see from the commit history that you run into some issues with linking statically. Maybe someone with more experience with linking C libraries in haskell can help, @RyanGlScott?

m-yac commented:
> I'm late with this feedback, but I think LLDB should be linked statically, not dynamically.

For some reason I got the impression from reading the docs around `lmdb-simple` that this couldn't be done, but doing some basic research now it seems like maybe we could in fact statically link the LMDB `.a` files and have `lmdb-simple` just work?

That being said, I don't have very much time left to work on this task, and something like this doesn't seem super easy to configure and get working (especially on CI). If Ryan comes back and says it's super easy then I can try to get it working, otherwise someone else might have to jump in. But, if someone just gets `lmdb-simple` working with a statically-linked LMDB on the CI, I can definitely do the rest (i.e. updating the solver caching code to use `lmdb-simple` again).

RyanGlScott commented:
Sorry, I'm missing context on the linking issues you're alluding to. Can you describe what the problem is in more detail?

m-yac commented:
> Sorry, I'm missing context on the linking issues you're alluding to. Can you describe what the problem is in more detail?

@RyanGlScott Originally, this PR used the [`lmdb-simple` Haskell library](https://hackage.haskell.org/package/lmdb-simple-0.4.0.0) for interacting with LMDB databases instead of calling out to Python. However, by default, this requires the dynamically-linked LMDB C library to be installed at runtime, the installation of which is too much to ask of every SAW user. The question: Is it possible, and if so, how easy would it be to statically link the LMDB C library when building SAW and have that work with `lmdb-simple`?

m-yac commented on src/SAWScript/Interpreter.hs:
```diff
@@ -452,6 +456,12 @@ buildTopLevelEnv proxy opts =
        ss <- basic_ss sc
        jcb <- JCB.loadCodebase (jarList opts) (classPath opts) (javaBinDirs opts)
        currDir <- getCurrentDirectory
+       mb_cache <- lookupEnv "SAW_SOLVER_CACHE_PATH" >>= \case
```
Added to the manual! I also added a section about solver caching which explains the feature in more detail.

RyanGlScott commented:
I still feel like I'm missing something. After all, the approach you are using still requires users to install LMDB themselves, but via Python instead of Haskell. I'm not sure why one approach is considered too much while the other isn't...

RyanGlScott commented:
OK. If I understand correctly, the issue isn't really in static vs. dynamic linking, but rather requiring users to install a third-party C library that powers an optional feature in SAW. From this perspective, I don't see that great of a difference between options (1) and (3), as statically linking against `lmdb` would still require users to install the C library separately on their system.

Unfortunately, this sort of thing is awkward to accomplish. I can foresee two ways to make this not terrible for SAW users, but both of them would require downstream changes to the Haskell `lmdb` bindings (or creating our own):

1. Have the Haskell `lmdb` library bundle a copy of the `lmdb` C library as part its sources, which are compiled when building `lmdb` with `cabal`. (This is basically the same approach that the Python `lmdb` bindings use, if I understand correctly.)
2. Use the same trick that Haskell's `GLUT` bindings use: don't link directly against `lmdb`, but instead call out to `lmdb` at runtime using `dlopen`/`LoadLibrary`. This would mean that if a user doesn't have `lmdb` installed when building SAW, everything would still work unless they explicitly enable caching.

Of these options, (1) would probably be easiest, considering that the Python bindings already do this.

m-yac commented:
Ah, okay – thanks, Ryan! For some reason I thought static linking would work similar to (1), i.e. that the C parts would be compiled into the `saw` executable, and the user wouldn't have to install anything. (Shows you how much I've actually used C/C++ in my life!). Sorry if that misunderstanding was part of your confusion.

Given the amount of time I have left to work on this, I think I'd prefer to just keep the current Python solution instead of trying to write a new Haskell LMDB library. @andreistefanescu what do you think?

On another note: I think I've addressed all of your comments now, @RyanGlScott! The only things left that I think need review are the new docs I added in 0e2efb201853daa14571ffd08f82d995b9ac632a and bbdfdf0eb863dc6148e9c12dbdff11d746a1ba9d, as well as the question I asked about running `pip install lmdb` on the CI.

RyanGlScott commented on src/SAWScript/Interpreter.hs:
```diff
@@ -1057,6 +1090,61 @@ primitives = Map.fromList
+  , prim "print_solver_cache_stats" "TopLevel ()"
+    (pureVal (onSolverCache printSolverCacheStats))
+    Experimental
+    [ "Print out statistics about how the solver cache has been used, namely"
+    , "how many entries are in the cache, whether the cache is being stored"
+    , "in memory or on disk, how many insertions into the cache have been made"
+    , "so far this session, how many failed insertion attempts have been made"
+    , "so far this session, how times cached results have been used so far this"
+    , "session, and with how many failed attempted usages have occurred so far"
+    , "this session." ]
```
I'd find it helpful to put each of these six things in a bulleted list:

```
    [ "Print out statistics about how the solver cache has been used, namely:"
    , "* How many entries are in the cache"
    , "* Whether the cache is being stored"
    , ...
    ]
```

RyanGlScott commented on src/SAWScript/Interpreter.hs:
```diff
@@ -1057,6 +1090,61 @@ primitives = Map.fromList
+  , prim "test_solver_cache_stats" "Int -> Bool -> Int -> Int -> Int -> Int -> TopLevel ()"
+    (pureVal test_solver_cache_stats)
+    Experimental
+    [ "Test whether the values of the statistics printed out by"
+    , "print_solver_cache_stats are equal to those given, failing if"
+    , "this does not hold. Specifically, the arguments represent how many"
+    , "entries are in the cache, whether the cache is being stored on disk,"
+    , "how many insertions into the cache have been made, how many failed"
+    , "insertion attempts have been made, how times cached results have"
+    , "been used, and how many failed attempted usages have occurred." ]
```
Similarly, use a bulleted list here.

RyanGlScott reviewed (commented):
> For some reason I thought static linking would work similar to (1), i.e. that the C parts would be compiled into the `saw` executable

To be clear, that is how static linking works. The contentious bit is where the static _library_ comes from, i.e., whether the user has to install it themselves, or if they can use a copy of the library that is bundled with `lmdb` itself.

In any case, I'm fine with using the Python approach. If you want, you could file an issue to track the idea of switching to an in-Haskell solution.

m-yac commented:
Okay, this is ready for review again! Here's what's been done:
- Update to use Galois' newly forked [Haskell LMDB bindings](https://github.com/galoisInc/lmdb) (thanks, @RyanGlScott!) which now statically link with the LMDB C library. This replaces the previous messy workaround of calling out to a Python process at runtime.
- Remove the ability to cache results purely in memory. It seemed like this feature was not very useful in the first place, and it would've required extra work to keep around now that we're using the Haskell LMDB bindings as our backend. This change means `enable_solver_cache` was removed and now the only ways to use solver caching are through the `SAW_SOVLER_CACHE_PATH` environment variable and the `set_solver_cache_path` command
- When using the `SAW_SOVLER_CACHE_PATH` environment variable, don't create an LMDB database at the given path until the cache is actually used (as requested by @weaversa)
- Save the timestamp of the last time each entry was used (as requested by @weaversa)
- Move the Python interface for interacting with the solver cache to the SAW Python API ([`solver_cache.py`](https://github.com/GaloisInc/saw-script/blob/c28e3405fefeb7465a85e68d90cce95d614548d3/saw-remote-api/python/saw_client/solver_cache.py)) since that interface is quite useful and should be made accessible somewhere. Unlike the set of solver caching commands available in `saw-script`, this interface can be used to perform arbitrary database operations on the solver cache – for example, deleting all entries with timestamps older than a specific date. **Note**: This `solver_cache.py` file has not yet been integrated with the rest of the SAW Remote API. We do intend to integrate it, and add the solver caching commands present in `saw-script` to the Remote API, but we're currently thinking that this will be done in a separate PR.

RyanGlScott commented on saw-remote-api/python/saw_client/solver_cache.py:
```diff
@@ -0,0 +1,124 @@
+import cbor2
+from collections.abc import MutableMapping
+from dataclasses import dataclass
+from datetime import datetime
+import dateutil.parser
+import lmdb
```
The remote API tests are failing in CI in part because the `cbor2` and `lmdb` libraries aren't declared in the `pyproject.toml` configuration.

RyanGlScott commented on intTests/test_solver_cache/test.sh:
```diff
@@ -0,0 +1,28 @@
+$SAW test_path_and_reuse.saw
+
+# Testing cleaning the solver cache
+pip install cbor2 python-dateutil lmdb
+python3 -i ../../saw-remote-api/python/saw_client/solver_cache.py << END
```
Does this test still need to rely on Python to work, or can it be implemented in terms of SAWScript? If it _does_ rely on Python, perhaps we should move this test to the test suite for the Python bindings.

m-yac replied:
Good point! It does not rely on the remote API – the only reason this bit calls out to Python is so that we can manually mess with the version strings of certain entries to make sure the `clean_solver_cache` command works. I removed the dependency on the remote API (instead just using the Python LMDB bindings directly) and added a comment explaining what this bit of Python code does.

RyanGlScott commented on src/SAWScript/SolverCache.hs:
```diff
@@ -0,0 +1,632 @@
+
+-- Orphan Instances and Helper Functions ---------------------------------------
+
+deriving instance Generic FirstOrderType
+deriving instance Generic FirstOrderValue
```
Do these instances need to be orphans, or could they instead be moved to `Verifier.SAW.FiniteValue`, where these data types are defined? I'm hoping it's the latter, since that would avoid needing to use `-fno-warn-orphans`.

m-yac replied:
Does your suggestion include the `ToJSON`/`FromJSON` instances as well? If not, `-fno-warn-orphans` would still need to be used for this file.

RyanGlScott replied:
Yes, the `{To,From}JSON` instances would also need to be moved to avoid them being orphans.

RyanGlScott commented on src/SAWScript/SolverCache.hs:
I find this code hard to understand since there are multiple layers of `Maybe`s involved. There are three possibilities—`Nothing`, `Just Nothing`, and `Just (Just x)`—and it's unclear to me what the distinction between the three are. Why do we pick `Just Nothing` here instead of `Nothing`, for instance?

At the very least, we should document the conventions at play. It might also be worth turning `Maybe SolverCacheValue` into a bespoke data type to make it more obvious what each of the states mean.

m-yac replied:
It think the problem was not the `Maybe SolverCacheValue`, but in fact the way I defined `SolverCacheOp`. I changed the definition of `SolverCacheOp` in bc446cf0ce29ad095e44d702ea6264e409337095 to remove the extra `Maybe` and also make things hopefully a lot more clear. This relevant bits are:
```haskell
-- | An operation on a 'SolverCache', returning a value of the given type as
-- well as an updated 'SolverCache' ('solverCacheOp'). Additionally, in the case
-- of no enabled solver cache, the operation could either fail or return a
-- default value ('solverCacheOpDefault').
data SolverCacheOp a = SCOpOrFail (Options -> SolverCache -> IO (a, SolverCache))
                     | SCOpOrDefault a (Options -> SolverCache -> IO (a, SolverCache))

-- | Get the operation associated to a 'SolverCacheOp'
solverCacheOp :: SolverCacheOp a -> Options -> SolverCache -> IO (a, SolverCache)
...

-- | Get the default value associated to a 'SolverCacheOp', if any
solverCacheOpDefault :: SolverCacheOp a -> Maybe a
...

-- | Lookup a 'SolverCacheKey' in the solver result cache
lookupInSolverCache :: SolverCacheKey -> SolverCacheOp (Maybe SolverCacheValue)
lookupInSolverCache k = SCOpOrDefault Nothing $ \opts cache@SolverCache{..} ->
...
```
Does that help make this less confusing?

RyanGlScott replied:
Much better! My brain started to make the Windows 95 shutdown noise when I read "`Just Nothing`", and I'm much happier now that that isn't there :)

RyanGlScott reviewed (commented):
There are some new test failures in the integration tests CI job. I've taken another pass through `SAWScript.SolverCache` now that it is based on `lmdb-simple`.

m-yac commented on src/SAWScript/SolverCache.hs:
```diff
@@ -0,0 +1,632 @@
+  where dropFO ('F':'O':tv:cs) | tv `elem` ['T', 'V'] = cs
+        dropFO cs = cs
+
+instance FromJSON FirstOrderType where
+  parseJSON = JSON.genericParseJSON firstOrderJSONOptions
+instance FromJSON FirstOrderValue where
+  parseJSON = JSON.genericParseJSON firstOrderJSONOptions
+
+instance ToJSON FirstOrderType where
+  toJSON = JSON.genericToJSON firstOrderJSONOptions
+  toEncoding = JSON.genericToEncoding firstOrderJSONOptions
+instance ToJSON FirstOrderValue where
+  toJSON = JSON.genericToJSON firstOrderJSONOptions
+  toEncoding = JSON.genericToEncoding firstOrderJSONOptions
```
I generally agree with you, but in this case the fact that I used [`firstOrderJSONOptions`](https://github.com/GaloisInc/saw-script/blob/9c5c3949c92d5fb6115bf1760ac2cf4eff3c4bf8/src/SAWScript/SolverCache.hs#L132) instead of `defaultJSONOptions` means it's not quite as brittle as it could be. By default, each constructor would be encoded by an integer, thus making it brittle with respect to constructor reordering and new constructors being added in the middle of others. However, `firstOrderJSONOptions` specifically encodes constructors like `FOWord` as `"Word"`. This means this encoding is independent of constructor order and new constructors being added. As long as no constructor names are changed, out-of-date encodings should always work.

As for the `saw-script` `git` commit, that is only included in the case of `rme` being the backend, since that sub-component has no versioning system unto itself. So, if a constructor name is ever changed, it is possible that `saw-script` will attempt to use an out-of-date encoding when retrieving a non-`rme` cached entry.

Regardless, I definitely agree that this assumption (that constructor names will not change) should be documented. If your comment about moving instances to `FiniteValue.hs` include these `ToJSON`/`FromJSON` instances as well, I will put the comment in `FiniteValue.hs` as well.

m-yac replied:
Added some docs (and resolved the comment above) in a960c3665db04562d30b9160682bcef359cb5fff!

RyanGlScott reviewed (approved):
Excellent, I'm quite happy with how this has turned out. Great work, @myac!
